src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -0,0 +1,1349 @@
+    // Return Value:
+    // - <none>
+    void ControlCore::_refreshSizeUnderLock()
+    {
+        const auto widthInPixels = _panelWidth * _compositionScaleX;
+        const auto heightInPixels = _panelHeight * _compositionScaleY;
+
+        _doResizeUnderLock(widthInPixels, heightInPixels);
+    }
+
+    // Method Description:
+    // - Process a resize event that was initiated by the user. This can either
+    //   be due to the user resizing the window (causing the swapchain to
+    //   resize) or due to the DPI changing (causing us to need to resize the
+    //   buffer to match)
+    // Arguments:
+    // - newWidth: the new width of the swapchain, in pixels.
+    // - newHeight: the new height of the swapchain, in pixels.
+    void ControlCore::_doResizeUnderLock(const double newWidth,
+                                         const double newHeight)
+    {
+        SIZE size;
+        size.cx = static_cast<long>(newWidth);
+        size.cy = static_cast<long>(newHeight);
+
+        // Don't actually resize so small that a single character wouldn't fit
+        // in either dimension. The buffer really doesn't like being size 0.
+        if (size.cx < _actualFont.GetSize().X || size.cy < _actualFont.GetSize().Y)
+        {
+            return;
+        }
+
+        _terminal->ClearSelection();
+
+        // Tell the dx engine that our window is now the new size.
+        THROW_IF_FAILED(_renderEngine->SetWindowSize(size));
+
+        // Invalidate everything
+        _renderer->TriggerRedrawAll();
+
+        // Convert our new dimensions to characters
+        const auto viewInPixels = Viewport::FromDimensions({ 0, 0 },
+                                                           { static_cast<short>(size.cx), static_cast<short>(size.cy) });
+        const auto vp = _renderEngine->GetViewportInCharacters(viewInPixels);
+
+        // If this function succeeds with S_FALSE, then the terminal didn't
+        // actually change size. No need to notify the connection of this no-op.
+        const HRESULT hr = _terminal->UserResize({ vp.Width(), vp.Height() });
+        if (SUCCEEDED(hr) && hr != S_FALSE)
+        {
+            _connection.Resize(vp.Height(), vp.Width());
+        }
+    }
+
+    void ControlCore::SizeChanged(const double width,
+                                  const double height)
+    {
+        _panelWidth = width;
+        _panelHeight = height;
+
+        auto lock = _terminal->LockForWriting();
+        const auto currentEngineScale = _renderEngine->GetScaling();
+
+        auto scaledWidth = width * currentEngineScale;
+        auto scaledHeight = height * currentEngineScale;
+        _doResizeUnderLock(scaledWidth, scaledHeight);
+    }
+
+    void ControlCore::ScaleChanged(const double scaleX,
+                                   const double scaleY)
+    {
+        if (!_renderEngine)
+        {
+            return;
+        }
+
+        _compositionScaleX = scaleX;
+        _compositionScaleY = scaleY;
+

```

Comment:
is it a worry that we're changing these without the lock?

Reply:
That's reasonable, moved inside lock